engine/src/main/java/org/pentaho/di/job/entries/sftp/SFTPClient.java

```diff
@@ -0,0 +1,479 @@
+      throw new KettleJobException( e );
+    } finally {
+      if ( inputStream != null ) {
+        try {
+          inputStream.close();
+        } catch ( IOException e ) {
+          throw new KettleJobException( e );
+        }
+      }
+    }
+  }
+
+  public void put( InputStream inputStream, String remoteFile ) throws KettleJobException {
+    int mode = ChannelSftp.OVERWRITE;
+
+    try {
+      c.put( inputStream, remoteFile, null, mode );
+    } catch ( Exception e ) {
+      throw new KettleJobException( e );
+    } finally {
+      if ( inputStream != null ) {
+        try {
+          inputStream.close();
+        } catch ( IOException e ) {
+          throw new KettleJobException( e );
+        }
+      }
+    }
+  }
+
+  public void delete( String file ) throws KettleJobException {
+    try {
+      c.rm( file );
+    } catch ( SftpException e ) {
+      throw new KettleJobException( e );
+    }
+  }
+
+  /**
+   * Creates the given folder. The {@param path} can be either absolute or relative.
+   * Allows creation of nested folders.
+   */
+  public void createFolder( String path ) throws KettleJobException {
+    try {
+      String[] folders = path.split( "/" );
+      folders[ 0 ] = ( path.charAt( 0 ) != '/' ? pwd() + "/" : "" ) + folders[ 0 ];
+
+      for ( int i = 1; i < folders.length; i++ ) {
+        folders[ i ] = folders[ i - 1 ] + "/" + folders[ i ];
+      }
+
+      for ( String f : folders ) {
+        if ( f.length() != 0 && !folderExists( f ) ) {
+          c.mkdir( f );
+        }
+      }
+    } catch ( ArrayIndexOutOfBoundsException e ) {
+      throw new KettleJobException( e );
+    } catch ( SftpException e ) {
+      throw new KettleJobException( e );
+    }
+  }
+
+  /**
+   * Rename the file.
+   */
+  public void renameFile( String sourcefilename, String destinationfilename ) throws KettleJobException {
+    try {
+      c.rename( sourcefilename, destinationfilename );
+    } catch ( SftpException e ) {
+      throw new KettleJobException( e );
+    }
+  }
+
+  public FileType getFileType( String filename ) throws KettleJobException {
+    try {
+      SftpATTRS attrs = c.stat( filename );
+      if ( attrs == null ) {
+        return FileType.IMAGINARY;
+      }
+
+      if ( ( attrs.getFlags() & SftpATTRS.SSH_FILEXFER_ATTR_PERMISSIONS ) == 0 ) {
+        throw new KettleJobException( "Unknown permissions error" );
+      }
+
+      if ( attrs.isDir() ) {
+        return FileType.FOLDER;
+      } else {
+        return FileType.FILE;
+      }
+    } catch ( Exception e ) {
+      throw new KettleJobException( e );
+    }
+  }
+
+  public boolean folderExists( String foldername ) {
+    boolean retval = false;
+    try {
+      SftpATTRS attrs = c.stat( foldername );
+      if ( attrs == null ) {
+        return false;
+      }
+
+      if ( ( attrs.getFlags() & SftpATTRS.SSH_FILEXFER_ATTR_PERMISSIONS ) == 0 ) {
+        throw new KettleJobException( "Unknown permissions error" );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [KettleJobException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [387].  Consider adding a catch clause of this specific exception in the outer try statement.